src/core/qgstemporalnavigationobject.cpp

```diff
@@ -0,0 +1,175 @@
+    email                : samweli at kartoza dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgstemporalnavigationobject.h"
+#include "qgis.h"
+
+QgsTemporalNavigationObject::QgsTemporalNavigationObject( QObject *parent )
+  : QObject( parent )
+{
+  mNewFrameTimer = new QTimer( this );
+
+  connect( mNewFrameTimer, &QTimer::timeout,
+           this,  qgis::overload<>::of( &QgsTemporalNavigationObject::timerTimeout ) );
+}
+
+
+void QgsTemporalNavigationObject::timerTimeout()
+{
+  if ( mPlayBackMode == PlaybackMode::Forward )
+    forward();
+  if ( mPlayBackMode == PlaybackMode::Reverse )
+    backward();
+  return;
+}
+
+QgsDateTimeRange QgsTemporalNavigationObject::dateTimeRangeForFrameNumber( long long frame ) const
+{
+  QDateTime start = mTemporalExtents.begin();
+
+  long long pastFrame = frame - 1;
+  if ( pastFrame < 0 )
+    pastFrame = 0;
+
+  QDateTime begin = start.addSecs( pastFrame * mFrameDuration.seconds() );
+  QDateTime end = start.addSecs( frame * mFrameDuration.seconds() );
+
+  if ( end <= mTemporalExtents.end() )
+    return QgsDateTimeRange( begin, end );
+
+  return mTemporalExtents;
+}
+
+void QgsTemporalNavigationObject::setTemporalExtents( QgsDateTimeRange temporalExtents )
+{
+  mTemporalExtents = temporalExtents;
+}
+
+QgsDateTimeRange QgsTemporalNavigationObject::temporalExtents() const
+{
+  return mTemporalExtents;
+}
+
+void QgsTemporalNavigationObject::setCurrentFrameNumber( long long frameNumber )
+{
+  if ( frameNumber < 0 ||
+       frameNumber >= totalFrameCount() )
+  {
+    pause();
+    return;
+  }
+
+  if ( mCurrentFrameNumber != frameNumber )
+  {
+    mCurrentFrameNumber = frameNumber;
+    QgsDateTimeRange range = dateTimeRangeForFrameNumber( mCurrentFrameNumber );
+    emit updateTemporalRange( range );
+  }
+}
+
+long long QgsTemporalNavigationObject::currentFrameNumber() const
+{
+  return mCurrentFrameNumber;
+}
+
+void QgsTemporalNavigationObject::setFrameDuration( QgsInterval frameDuration )
+{
+  mFrameDuration = frameDuration;
+}
+
+QgsInterval QgsTemporalNavigationObject::frameDuration() const
+{
+  return mFrameDuration;
+}
+
+void QgsTemporalNavigationObject::setFramesPerSeconds( double framesPerSeconds )
+{
+  if ( framesPerSeconds > 0 )
+    mFramesPerSecond = framesPerSeconds;
```

Comment:
mNewFrameTimer->setInterval( ( 1.0 / mFramesPerSecond ) * 1000 );

(will update the frame rate for currently playing animations too)

Reply:
this will be handled in the play slot, which is calling mNewFrameTimer->start( ( 1.0 / mFramesPerSecond ) * 1000 );

Reply:
Yes, but if the animation is already playing, then you'd need to stop and restart it for that slot to be called. That's why we also need to update the timeout here -- a playing animation will instantly respond to the frame rate change

Reply:
The current logic already handles that.

Reply:
Where? there's no code here which updates the timer.

Reply:
Yes, you are right, didn't get your point at first.